File: templates/problems/15_greedy/04_gas_station/python/solution.py
def solve_can_complete_circuit(gas: list[int], cost: list[int]) -> int:
    """Find the starting gas station for a circular route. O(n) time, O(1) space."""
    total_surplus = 0
    current_surplus = 0
    start = 0
    for i in range(len(gas)):
        total_surplus += gas[i] - cost[i]
        current_surplus += gas[i] - cost[i]
        if current_surplus < 0:
            start = i + 1
            current_surplus = 0
    if total_surplus < 0:
        return -1
    return start

File: templates/problems/15_greedy/04_gas_station/typescript/solution.ts
/**
 * Find the starting gas station for a circular route. O(n) time, O(1) space.
 */
export function solveCanCompleteCircuit(gas: number[], cost: number[]): number {
    let totalSurplus = 0;
    let currentSurplus = 0;
    let start = 0;
    for (let i = 0; i < gas.length; i++) {
        totalSurplus += gas[i] - cost[i];
        currentSurplus += gas[i] - cost[i];
        if (currentSurplus < 0) {
            start = i + 1;
            currentSurplus = 0;
        }
    }
    if (totalSurplus < 0) return -1;
    return start;
}
